compiler/mir/mirgen.nim

```diff
@@ -0,0 +1,2105 @@
+    ## No further operators can be chained after an operator marked as
+    ## ``ChainEnd``
+
+  Block = object
+    ## Information about a ``block``
+    label: PSym ## the symbol of the block's label. 'nil' if the block has no
+                ## label
+    id: LabelId ## the block's internal label ID
+
+  CodeFragment = object
+    nodes: MirNodeSeq
+    source: seq[SourceIndex]
+
+  SourceProvider = object
+    ## Stores the active origin and the in-progress database of origin
+    ## ``PNode``s. Both are needed together in most cases, hence their bundling
+    ## into an object
+    active: tuple[n: PNode, id: SourceId]
+      ## the ``PNode`` to use as the origin for emitted ``MirNode``s (if none
+      ## is explicitly provided). If `id` is 'none', no database entry exists
+      ## for the ``PNode`` yet
+    store: Store[SourceIndex, PNode]
+      ## the in-progress database of origin ``PNode``s
+
+  GenOption* = enum
+    goIsNimvm     ## choose the ``nimvm`` branch for ``when nimvm`` statements
+    goGenTypeExpr ## don't omit type expressions
+
+  TCtx = object
+    # output:
+    stmts: CodeFragment
+
+    # working state:
+    staging: CodeFragment ## intermediate buffer for partially generated MIR
+                          ## code
+
+    blocks: seq[Block] ## the stack of active ``block``s. Used for looking up
+                       ## break targets
+
+    sp: SourceProvider
+
+    tmpMap: Table[ItemId, TempId]
+      ## maps the ID of ``skTemp`` symbols to the ``TempId`` used to refer to
+      ## them in output
+
+    numTemps: int  ## provides the ID for temporaries
+    numLabels: int ## provides the ID to use for the next label
+
+    # input:
+    context: TSymKind ## what entity the input AST is part of (e.g. procedure,
+                      ## macro, module, etc.). Used to allow or change how the
+                      ## AST is interpreted in some places
+    graph: ModuleGraph
+
+    options: set[GenOption]
+
+const
+  abstractInstTypeClass = abstractInst + tyUserTypeClasses
+  # TODO: this set shouldn't be needed. ``tyUserTypeClass`` and
+  #       ``tyUserTypeClassInst`` should be turned into either aliases or
+  #       ``tyGenericInst`` types when they're resolved
+  ComplexExprs = {nkIfExpr, nkCaseStmt, nkBlockExpr, nkTryStmt}
+    ## The expression that are treated as complex and which are transformed
+    ## into assignments-to-temporaries
+
+func len[T](x: HOslice[T]): int {.inline.} =
+  int(x.b - x.a)
+
+func halfOpen[T](x: Slice[T]): HOslice[T] {.inline.} =
+  ## Constructs a half-open slice from a normal slice. `x` is expected to
+  ## already represent a half-open slice
+  HOslice[T](a: x.a, b: x.b)
+
+template toOpenArray[T, I](x: seq[T], slice: Slice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int)
+
+template toOpenArray[T, I](x: seq[T], slice: HOslice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int - 1)
```

Comment:
I highly recommend caching `slice` in a temporary to avoid multiple evaluations.